### DISCUSSION TITLE
Argument type for filter argument of custom struct

### DISCUSSION THREAD

**Discussion by jinschoi** [mitsuhiko/minijinja#335] [Q&A]
I'm using minijinja as the templating library for a small web app. I make a bunch of sqlx calls and instantiate them as custom serializable structs which are passed to minijinja. (I would love to just pass the anonymous struct objects from calls to the query! macro, but they are not serializable.) Since different places need different fields, these structs are often local to an endpoint and contain different subsets of fields from the same tables.

I would like to write a filter that takes a struct object that contains certain required fields and returns a String. Since these structs are all dynamic Value objects once inside a template, I thought I could do this fairly simply, but I cannot figure it out. I tried using ViaDeserialize, but that only works with one specific concrete type.

What should the argument type for my filter be? Can I just use Value and deserialize it into a custom struct that just contains the fields that I want to use?

**Comment by mitsuhiko** [comment]
That depends a bit on how the structs are passed there. Are they using `Object`? In that case you can use `downcast_object_ref`.

```rust
struct Query { ... }

impl Object for Query { ... }

/// Utility function to extract a [`Query`] out of a [`Value`].
fn value_as_query(obj: &Value) -> Result<&Query, Error> {
    obj.downcast_object_ref::<Query>().ok_or_else(|| {
        Error::new(
            ErrorKind::InvalidOperation,
            "filter must be applied to query",
        )
    })
}

/// Filters a query by some keyword arguments as filter function.
fn filter_filter(obj: &Value, kwargs: Kwargs) -> Result<Value, Error> {
    Ok(Value::from_object(value_as_query(obj)?.filter(kwargs)))
}
```

See this example for the full code: https://github.com/mitsuhiko/minijinja/blob/main/examples/dsl/src/main.rs

**Reply by jinschoi** [reply]
No, not using Object, just passing them as Serialize.

I think I may have figured out something that works for me that is like what I described in the question:
```rust
// A and B are Serializable results from sqlx queries that get passed to templates
struct A {
  a: i32,
  b: i32,
  c: i32,
}

struct B {
  b: i32,
  c: i32,
  d: i32,
}

// C is Deserialize, and just the subset of fields from A and B that our particular filter cares about:
struct C {
  b: i32,
  c: i32,
}

// my_filter just uses fields b and c and applies a common formatting operation on them.
fn my_filter(v: Value) -> Result<String, Error> {
  let c = C::deserialize(v)?;
  let res = format!("{} {}", c.b, c.c);
  Ok(res)
}

// That works with {{ a_or_b | my_filter }} from a template, but if I want to call it from code, I need to turn it into a Value first:
fn somefunc() {
  let a: A = ...;
  let formatted = my_filter(&Value::from_serializable(&a))?;
}
```
 
Is there a simpler, more idiomatic way of getting this behavior? I don't think using Object would help because downcast_ref just returns the boxed struct, and I have many different single shot structs that represent the type of individual sqlx queries, all with different subsets of fields.

**Reply by mitsuhiko** [reply]
You can directly use `ViaDeserialize`:

```rust
fn my_filter(v: ViaDeserialize<C>) -> String {
    format!("{} {}", c.b, c.c)
}
```

See also this example: https://github.com/mitsuhiko/minijinja/blob/eac92fa2c0f8b12277938469051e37ef627211b3/examples/deserialize/src/main.rs#L20-L22

**Reply by jinschoi** [reply]
But what do I pass it if I want to call it from code with a struct A or B, as in my somefunc() example?

**Reply by mitsuhiko** [reply]
You just wrap it in `ViaDeserialize`:

```rust
my_filter(ViaDeserialize(C { ... }))
```

**Reply by jinschoi** [reply]
Okay. That would require that I convert any A or B to C, either explicitly at the call site or through multiple From implementations. I think for my case, going through dynamic Values is easier.

**Reply by mitsuhiko** [reply]
@jinschoi why are you calling these functions to begin with? It's pretty rare that filters would be called from outside of the engine (ignoring tests).

**Reply by jinschoi** [reply]
I have a utility formatting filter, link_to_thing, which when given a "thing" struct with some particular fields, constructs an HTTP link from it using various fields to add status text to the link, etc. It is mainly to be used as a filter within templates, but sometimes when working with AJAX driven components, a JS table for example, that returns JSON with HTML snippets, it is necessary to call link_to_thing from within code.
